pkg/controller/periodic/job.go

```diff
@@ -108,19 +106,33 @@ func (j *JobController) CreateGathererJob(ctx context.Context, dataGatherName, i
 	return j.kubeClient.BatchV1().Jobs(insightsNamespace).Create(ctx, gj, metav1.CreateOptions{})
 }
 
-// WaitForJobCompletion polls the Kubernetes API every 20 seconds and checks if the job finished.
+// WaitForJobCompletion listen the Kubernetes events to check if job finished.
 func (j *JobController) WaitForJobCompletion(ctx context.Context, job *batchv1.Job) error {
-	return wait.PollUntilContextCancel(ctx, 20*time.Second, true, func(ctx context.Context) (done bool, err error) {
-		j, err := j.kubeClient.BatchV1().Jobs(insightsNamespace).Get(ctx, job.Name, metav1.GetOptions{})
-		if errors.IsNotFound(err) {
-			return false, err
-		}
-		if j.Status.Succeeded > 0 {
-			return true, nil
-		}
-		if j.Status.Failed > 0 {
-			return true, fmt.Errorf("job %s failed", job.Name)
+	watcher, err := j.kubeClient.BatchV1().Jobs(insightsNamespace).
+		Watch(ctx, metav1.ListOptions{FieldSelector: fmt.Sprintf("metadata.name=%s", job.Name)})
+	if err != nil {
+		return err
+	}
+	defer watcher.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case event, ok := <-watcher.ResultChan():
+			if !ok {
+				return fmt.Errorf("watcher channel was closed unexpectedly")
+			}
+			if event.Type == watch.Error {
+				return fmt.Errorf("watcher received error event: %v", event.Object)
+			}
+			job := event.Object.(*batchv1.Job)
```

Comment:
What happens when this fails please?

Reply:
Unfortunately this doesn't help. https://play.golang.com/p/l7mo2jQifAT

Reply:
OK. I was reading the stuff, and it would not be a problem, once the watcher is looking just for Jobs:

```go
j.kubeClient.BatchV1().Jobs(insightsNamespace).Watch(ctx, metav1.ListOptions{})
```

The `event.Object` would be always a `runtime.Object` that can be converted to `batchv1.Job`. So, there is no reason for us to worry about it.

Reply:
Yes I know it should not happen, but it's better to cover all possible scenarios than to be suddenly surprised in my opinion.